docs/pkg_create.3
.Dd December 12, 2025
.Dt PKG_CREATE 3
.Os
.Sh NAME
.Nm pkg_create , pkg_create_i ,
.Nm pkg_create_new , pkg_create_free ,
.Nm pkg_create_set_format , pkg_create_set_overwrite ,
.Nm pkg_create_set_compression_level , pkg_create_set_rootdir ,
.Nm pkg_create_set_output_dir , pkg_create_set_timestamp ,
.Nm pkg_create_set_compression_threads
.Nd create packages
.Sh LIBRARY
.Lb libpkg
.Sh SYNOPSIS
.In pkg.h
.Ft struct pkg_create *
.Fn pkg_create_new
.Ft void
.Fn pkg_create_free "struct pkg_create *"
.Ft bool
.Fn pkg_create_set_format "struct pkg_create *" "const char *"
.Ft void
.Fn pkg_create_set_compression_level "struct pkg_create *" "int"
.Ft void
.Fn pkg_create_set_compression_threads "struct pkg_create *" "int"
.Ft void
.Fn pkg_create_set_overwrite "struct pkg_create *" "bool"
.Ft void
.Fn pkg_create_set_rootdir "struct pkg_create *" "const char *"
.Ft void
.Fn pkg_create_set_output_dir "struct pkg_create *" "const char *"
.Ft void
.Fn pkg_create_set_timestamp "struct pkg_create *" "time_t"
.Ft int
.Fn pkg_create "struct pkg_create *" "const char *" "const char *" "bool"
.Ft int
.Fn pkg_create_i "struct pkg_create *" "struct pkg *" "bool"
.Sh DESCRIPTION
.Fn pkg_create_new
Allocates a new
.Ft struct pkg_create *
which should be freed by the caller using
.Fn pkg_create_free .
.Pp
.Fn pkg_create_set_format
will define the compression format to use.
By default
.Qq txz
except if specified otherwise in
.Xr pkg.conf 5 .
It returns
.Va false
if the compression format is unknown and will fall back to the default format.
.Pp
.Fn pkg_create_set_compression_level
takes an
.Ft int
argument which represents the expected compression level.
3 special values are accepted:
.Bl -tag -width indentation
.It Va INT_MIN
Set the fastest compression format value
.It Va 0
Set the default (as specified in libarchive)
.It Va INT_MAX
Set the best compression ratio
.El
.Pp
.Fn pkg_create_set_compression_threads
takes an
.Ft int
argument which represents the expected number of threads used during
compression.
.Pp
.Fn pkg_create_set_overwrite
Accepts a boolean to define the default behaviour when creating a package if
a local file already exists.
The default behaviour is to overwrite.
.Pp
.Fn pkg_create_set_rootdir
Tells the program where to find the root directory used for packaging (this directory
is also known as the staging area).
If none is provided pkg will consider
.Sq Va / .
.Pp
.Fn pkg_create_set_output_dir
Tells the program where to create the new packages.
By default it will consider the current directory of the running program.
.Pp
.Fn pkg_create_set_timestamp
Define the timestamp to use as the creation and modification time for the files to
be packaged.
By default it will use the timestamp of the files on the filesystem.
.Pp
.Fn pkg_create
create a package from scratch using the
.Va metadata
and
.Va plist
as input.
The
.Va metadata
can be either directly a UCL manifest, or a directory containing multiple files:
.Bl -tag -width +MANIFEST
.It +MANIFEST
The UCL manifest
.It +DESC
A textual description (optional)
.It shell scripts (optional):
.Bl -tag -width pkg-post-deinstall
.It +INSTALL
Shell script that will be run during both post and pre install (DEPRECATED)
.It +PRE_INSTALL
Shell script that will be run during pre install phase
.It +POST_INSTALL
Shell script that will be run during post install phase
.It +DEINSTALL
Shell script that will be run during both post and pre deinstall (DEPRECATED)
.It +PRE_DEINSTALL
Shell script that will be run during pre deinstall phase
.It +POST_DEINSTALL
Shell script that will be run during post deinstall phase
.It pkg-install
Shell script that will be run during both post and pre install (DEPRECATED)
.It pkg-pre-install
Shell script that will be run during pre install phase
.It pkg-post-install
Shell script that will be run during post install phase
.It pkg-deinstall
Shell script that will be run during both post and pre deinstall (DEPRECATED)
.It pkg-pre-deinstall
Shell script that will be run during pre deinstall phase
.It pkg-post-deinstall
Shell script that will be run during post deinstall phase
.El
.It lua scripts (optional):
.Bl -tag -width pkg-post-deinstall.lua
.It pkg-pre-install.lua
Lua script that will be run during pre install phase
.It pkg-post-install.lua
Lua script that will be run during post install phase
.It pkg-pre-deinstall.lua
Lua script that will be run during pre deinstall phase
.It pkg-post-deinstall.lua
Lua script that will be run during post deinstall phase
.El
.El
.Pp
.Fn pkg_create_i
create a package for the provided installed
.Ar pkg .
If
.Ar hash
is set to
.Va true
then the file name will be in the form:
<pkgname>-<pkgversion>-<shorthash>.<extension>
instead of
<pkgname>-<pkgversion>.<extension>.


NEWS
Changes from 2.7.4 to 2.7.5
- backup lib: loudly complain if a library is not in filesystem but in plist
- backup lib: fix with rootdir
- progress: avoid useless newlines
- config: accept more boolean types
- fix fflags support

Changes from 2.7.3 to 2.7.4
- query: fix %y/%Y
- repo: do not delete repositories in overrides, simply disable them
- Declare CheriBSD and CheriABI OS and ABI

Changes from 2.7.2 to 2.7.3
- update: plug memory leak
- upgrade: plug leak
- upgrade: don't propose to downgrade to solve a provide/require

Changes from 2.7.1 to 2.7.2
- triggers: make the perpackage trigger per file instead of per directory
- triggers: fix execution of per packages triggers on pkg add/register
- fetch: respect RFC 3986

Changes from 2.7.0 to 2.7.1
- install: respect -q
- version: add IGNORE_MAJOR to pkg.conf and document it
- vital: now explicitly state the SAT solver issue are due to a vital package
- pkg_config: accept int as boolean converted via != 0
- lock: do not error in expected (un)lock state
- backup_libraries: now fully respects rootdir
- backup libraries: atomic replacement of the library
- Honor --repository flag for disabled repos as documented
- DB: loudly complain if the DB cannot be opened
- lua: add pkg.exec_capture, a pkg.exec version capturing stdout/stderr
- shlibs: also remove ignored shlibs provided from shlibs required

Changes from 2.6.99.6 to 2.7.0
- search: fix handling when search returns multiple JSON formatted results
- fix overlinking on macOS
- pkg.8: use % shell prompt for normal user

Changes from 2.6.99.5 to 2.6.99.6
- query: fix empty line output for purely multiline formats breaking poudriere
- pkg-backup: make the compression configurable
- sign_pkg.sh: use the .pkg extension

Changes from 2.6.99.4 to 2.6.99.5
- libpkg: fix --register-only with empty packages
- query: fix crash on multiline

Changes from 2.6.99.3 to 2.6.99.4
- info: fix version comparison with multi-character operators (>=, <=)
- fix --register-only to populate config file contents
- attempt config file merge for force install
- query: add %y and %Y for provides/requires

Changes from 2.6.99.2 to 2.6.99.3
- libfetch: fix segfault at exit when using SRV mirror type
- solver: avoid spurious reinstall from wrong repo with multi-repo setups

Changes from 2.6.99.1 to 2.6.99.2
- fetch: switch back to libfetch, completely remove curl
- add: resolve shlibs and provides via symlink directory layout
- add: ignore dead symlinks for provides/requires
- add: skip osversion check when force is set
- install -A: do not mark upgraded/reinstalled packages as automatic
- DB: fix a race condition in db initialization
- 400.status-pkg: provide an option to select the source to check
- docs: all manpages now mention -r can be called multiple times
- docs: update usage for create, install and set commands

Changes from 2.6.99.0 to 2.6.99.1
- greatly improve the test coverage
- add support for trigger per package (pre installation, post installation)
- fix info export wrong json for multiple packages
- solver: many fixes preventing useless extra reinstallation
- jobs: prevent install -f from deleting packages
- upgrade: prevent rdeps discovery explosion
- fix crash when using external merge tool
- fix ctrl-c during download not really killing pkg
- lots of code cleanup
- prevent running in jail if fflages are involved and the jail does not allow them
- prevent running if secure_level will block them
- improve performance by adding new indexes
- don't try anymore to upgrade the DB if it is readonly
- only open the db RW if needed
- fix pkg -N with WAL
- repositories: add per repo ssh_args
- delete,install,upgrade: add --autoremove
- pkg install foo if foo is installed will now mark it as non automatic
- pkg install -A foo if foo is installed will now mark it as automatic
- query: print an empty line instead of nothing if querying a multiline with no value
- install: accept https:// as an argument
- plenty of small fixes
- updating support globs
- delete: fix -fR
- search: add -Scomment-description
- lots of documentation updates

Changes from 2.5.99.3 to 2.6.99.0
- triggers: fix typo regex -> regexp
- remove FreeBSD 9 compat code
- lots of documentation behaviour
- fix ordering of pkg autoremove

Changes from 2.5.99.2 to 2.5.99.3
- extract: revert the extact optimization:
  - on low ended hardware it may be actually worst
  - it has too many side effects

Changes from 2.5.99.1 to 2.5.99.2
- pkgdb: skip WAL journal mode on read-only databases
- pkgdb: open read-only databases in immutable mode

Changes from 2.5.99.0 to 2.5.99.1
- Improve debug messages
- create: use -t to specify the MANIFEST file timestamps
- @dir: fix permission set via @mode
- manpages: fix some spelling
- progress:
  * pad the counters to improve the output,
  * remove the hash from the package names
- update mumhash to latest version
- fetch: add -s/--sumlnk option for mirror mode.
- fetch: honor PKG_REPO_SYMLINK env var
- triggers: only defer triggers which will fork/exec something
- db: enable WAL for databases on local file systems
- query: add support for evaluation of complex attributes
- store ignored shlibs in manifest and db
- Allow to define shlibs_{proviced,required}_ignore in the manifest directly
- configure: use some C23 features to reduce the number of checks in the configure
  script.
- drastically reduce memory usage
- fix an infinite loop on pkg autoremove
- optimization: do not extract files if the installed oone is identical
- if an extractions fails "file null mounted" complain and leave a file named
  .pkgnew aside
- upgrade multiple fixes on decision making which might have result in usesless
  request to reinstall
- pkg add ./file now check the version of the installed package if any and proposes
  to upgrade if necessary
- fix a regression in url_decode
- fix reproducibility issue with the packages
- fix a crash in pkg upgrade
- fix a crash when ftruncate() fails

Changes from 2.4.99.2 to 2.5.99.0

libpkg/pkg_create.c
/*-
 * Copyright (c) 2011-2020 Baptiste Daroussin <bapt@FreeBSD.org>
 * Copyright (c) 2011-2012 Julien Laffaye <jlaffaye@FreeBSD.org>
 * Copyright (c) 2014-2015 Matthew Seaman <matthew@FreeBSD.org>
 * Copyright (c) 2014 Vsevolod Stakhov <vsevolod@FreeBSD.org>
 * Copyright (c) 2023 Serenity Cyber Security, LLC
 *                    Author: Gleb Popov <arrowd@FreeBSD.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer
 *    in this position and unchanged.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/stat.h>

#include <errno.h>
#include <regex.h>
#include <fcntl.h>

#include <bsd_compat.h>

#include "pkg.h"
#include "private/event.h"
#include "private/pkg.h"
#include "private/pkg_abi.h"
#include "xmalloc.h"

#define TICK	100

static int load_metadata(struct pkg *pkg, const char *metadata, const char *plist,
    const char *rootdir);
static void fixup_abi(struct pkg *pkg, const char *rootdir, bool testing);
static void counter_init(const char *what, int64_t max);
static void counter_count(void);
static void counter_end(void);

extern struct pkg_ctx ctx;

static int
pkg_create_from_dir(struct pkg *pkg, const char *root,
		    struct pkg_create *pc, struct packing *pkg_archive,
		    bool trust_filesystem)
{
	char		 fpath[MAXPATHLEN];
	struct pkg_file	*file = NULL;
	struct pkg_dir	*dir = NULL;
	int		 ret;
	struct stat	 st;
	int64_t		 flatsize = 0;
	int64_t		 nfiles;
	const char	*relocation;
	char		*manifest;
	ucl_object_t	*obj;
	hardlinks_t	 hardlinks = vec_init();
	ssize_t		 linklen;

	if (pkg_is_valid(pkg) != EPKG_OK) {
		pkg_emit_error("the package is not valid");
		return (EPKG_FATAL);
	}

	relocation = pkg_kv_get(&pkg->annotations, "relocated");
	if (relocation == NULL)
		relocation = "";
	if (ctx.pkg_rootdir != NULL)
		relocation = ctx.pkg_rootdir;

	/*
	 * Get / compute size / checksum if not provided in the manifest
	 */

	nfiles = pkghash_count(pkg->filehash);
	counter_init("file sizes/checksums", nfiles);

	while (pkg_files(pkg, &file) == EPKG_OK) {

		snprintf(fpath, sizeof(fpath), "%s%s%s", root ? root : "",
		    relocation, file->path);

		if (lstat(fpath, &st) == -1) {
			pkg_emit_error("file '%s' is missing", fpath);
			vec_free_and_free(&hardlinks, free);
			return (EPKG_FATAL);
		}

		if (!(S_ISREG(st.st_mode) || S_ISLNK(st.st_mode))) {
			pkg_emit_error("file '%s' is not a regular file or symlink", fpath);
			vec_free_and_free(&hardlinks, free);
			return (EPKG_FATAL);
		}

		if (file->size == 0)
			file->size = (int64_t)st.st_size;

		if (st.st_nlink == 1 || !check_for_hardlink(&hardlinks, &st)) {
			flatsize += file->size;
		}

		if (file->perm == 0) {
			file->perm = st.st_mode & ~S_IFMT;
		}

		if (trust_filesystem) {
			free(file->sum);
			file->sum = pkg_checksum_generate_file(fpath,
							       PKG_HASH_TYPE_BLAKE2_BASE32);
			if (file->sum == NULL) {
				vec_free_and_free(&hardlinks, free);
				return (EPKG_FATAL);
			}

			if (S_ISLNK(st.st_mode)) {
				char link[MAXPATHLEN] = { 0 };
				linklen = readlink(fpath, link, sizeof(link) -1);
				if (linklen == -1) {
					vec_free_and_free(&hardlinks, free);
					pkg_emit_errno("pkg_create_from_dir", "readlink failed");
					return (EPKG_FATAL);
				}
				free(file->symlink_target);
				file->symlink_target = xstrdup(link);
			}

			if (pc->timestamp > (time_t)-1) {
				file->time[0].tv_sec = pc->timestamp;
				file->time[1].tv_sec = pc->timestamp;
			} else {
				file->time[0] = st.st_atim;
				file->time[1] = st.st_mtim;
			}
		}

		counter_count();
	}
	vec_free_and_free(&hardlinks, free);

	while (pkg_dirs(pkg, &dir) == EPKG_OK) {
		snprintf(fpath, sizeof(fpath), "%s%s%s", root ? root : "",
			 relocation, dir->path);

		if (lstat(fpath, &st) == -1) {
			pkg_emit_error("dir '%s' is missing", fpath);
			return (EPKG_FATAL);
		}

		if (!S_ISDIR(st.st_mode)) {
			pkg_emit_error("dir '%s' is not a directory", fpath);
			return (EPKG_FATAL);
		}

		if (dir->perm == 0) {
			dir->perm = st.st_mode & ~S_IFMT;
		}
	}

	counter_end();

	pkg->flatsize = flatsize;

	if (pkg->type == PKG_OLD_FILE) {
		pkg_emit_error("Cannot create an old format package");
		return (EPKG_FATAL);
	}

	obj = pkg_emit_object(pkg, PKG_MANIFEST_EMIT_COMPACT);
	manifest = ucl_object_emit(obj, UCL_EMIT_JSON_COMPACT);
	ucl_object_unref(obj);
	packing_append_buffer(pkg_archive, manifest, "+COMPACT_MANIFEST", strlen(manifest));
	free(manifest);
	obj = pkg_emit_object(pkg, 0);
	if (pc->expand_manifest) {
		manifest = ucl_object_emit(obj, UCL_EMIT_CONFIG);
	} else {
		manifest = ucl_object_emit(obj, UCL_EMIT_JSON_COMPACT);
	}
	ucl_object_unref(obj);
	packing_append_buffer(pkg_archive, manifest, "+MANIFEST", strlen(manifest));
	free(manifest);

	counter_init("packing files", nfiles);

	while (pkg_files(pkg, &file) == EPKG_OK) {
		char dpath[MAXPATHLEN];
		const char *dp = file->path;

		if (pkg->oprefix != NULL) {
			size_t l = strlen(pkg->prefix);
			if (strncmp(file->path, pkg->prefix, l) == 0 &&
			    (file->path[l] == '/' || l == 1)) {
				snprintf(dpath, sizeof(dpath), "%s%s%s",
				    pkg->oprefix, l == 1 ? "/" : "", file->path + l);
				dp = dpath;
			}
		}

		snprintf(fpath, sizeof(fpath), "%s%s%s", root ? root : "",
		    relocation, file->path);

		ret = packing_append_file_attr(pkg_archive, fpath, dp,
		    file->uname, file->gname, file->perm, file->fflags);
		if (ctx.developer_mode && ret != EPKG_OK)
			return (ret);
		counter_count();
	}

	counter_end();

	nfiles = pkghash_count(pkg->dirhash);
	counter_init("packing directories", nfiles);

	while (pkg_dirs(pkg, &dir) == EPKG_OK) {
		snprintf(fpath, sizeof(fpath), "%s%s%s", root ? root : "",
		    relocation, dir->path);

		ret = packing_append_file_attr(pkg_archive, fpath, dir->path,
		    dir->uname, dir->gname, dir->perm, dir->fflags);
		if (ctx.developer_mode && ret != EPKG_OK)
			return (ret);
		counter_count();
	}

	counter_end();

	return (EPKG_OK);
}

static struct packing *
pkg_create_archive(struct pkg *pkg, struct pkg_create *pc, unsigned required_flags)
{
	char		*pkg_path = NULL;
	struct packing	*pkg_archive = NULL;

	/*
	 * Ensure that we have all the information we need
	 */
	if (pkg->type != PKG_OLD_FILE)
		assert((pkg->flags & required_flags) == required_flags);

	if (pkg_mkdirs(pc->outdir) != EPKG_OK)
		return NULL;

	if (pkg_asprintf(&pkg_path, "%S/%n-%v", pc->outdir, pkg, pkg) == -1) {
		pkg_emit_errno("pkg_asprintf", "");
		return (NULL);
	}

	if (packing_init(&pkg_archive, pkg_path, pc->format,
	    pc->compression_level, pc->compression_threads, pc->timestamp, pc->overwrite) != EPKG_OK) {
		pkg_archive = NULL;
	}

	free(pkg_path);

	return pkg_archive;
}

static const char * const scripts[] = {
	"+INSTALL",
	"+PRE_INSTALL",
	"+POST_INSTALL",
	"+POST_INSTALL",
	"+DEINSTALL",
	"+PRE_DEINSTALL",
	"+POST_DEINSTALL",
	"pkg-install",
	"pkg-pre-install",
	"pkg-post-install",
	"pkg-deinstall",
	"pkg-pre-deinstall",
	"pkg-post-deinstall",
	NULL
};

static const char * const lua_scripts[] = {
	"pkg-pre-install.lua",
	"pkg-post-install.lua",
	"pkg-pre-deinstall.lua",
	"pkg-post-deinstall.lua",
	NULL
};

struct pkg_create *
pkg_create_new(void)
{
	struct pkg_create *pc;

	pc = xcalloc(1, sizeof(*pc));
	pc->format = packing_format_from_string(ctx.compression_format);
	pc->compression_level = ctx.compression_level;
	pc->compression_threads = ctx.compression_threads;
	pc->timestamp = (time_t) -1;
	pc->overwrite = true;
	pc->expand_manifest = false;

	return (pc);
}

void
pkg_create_free(struct pkg_create *pc)
{
	free(pc);
}

bool
pkg_create_set_format(struct pkg_create *pc, const char *format)
{
	if (STREQ(format, "tzst"))
		pc->format = TZS;
	else if (STREQ(format, "txz"))
		pc->format = TXZ;
	else if (STREQ(format, "tbz"))
		pc->format = TBZ;
	else if (STREQ(format, "tgz"))
		pc->format = TGZ;
	else if (STREQ(format, "tar"))
		pc->format = TAR;
	else
		return (false);
	return (true);
}

void
pkg_create_set_compression_level(struct pkg_create *pc, int clevel)
{
	pc->compression_level = clevel;
}

void
pkg_create_set_compression_threads(struct pkg_create *pc, int threads)
{
	pc->compression_threads = threads;
}

void
pkg_create_set_expand_manifest(struct pkg_create *pc, bool expand)
{
	pc->expand_manifest = expand;
}

void
pkg_create_set_rootdir(struct pkg_create *pc, const char *rootdir)
{
	pc->rootdir = rootdir;
}

void
pkg_create_set_output_dir(struct pkg_create *pc, const char *outdir)
{
	pc->outdir = outdir;
}

void
pkg_create_set_timestamp(struct pkg_create *pc, time_t timestamp)
{
	pc->timestamp = timestamp;
}

void
pkg_create_set_overwrite(struct pkg_create *pc, bool overwrite)
{
	pc->overwrite = overwrite;
}

static int
hash_file(struct pkg *pkg)
{
	char hash_dest[MAXPATHLEN];
	char filename[MAXPATHLEN];

	/* Find the hash and rename the file and create a symlink */
	pkg_snprintf(filename, sizeof(filename), "%n-%v.pkg",
			pkg, pkg);
	pkg->sum = pkg_checksum_generate_file(filename,
			PKG_HASH_TYPE_BLAKE2_BASE32);
	pkg_snprintf(hash_dest, sizeof(hash_dest), "%n-%v-%z.pkg",
			pkg, pkg, pkg);

	pkg_debug(1, "Rename the pkg file from: %s to: %s",
			filename, hash_dest);
	if (rename(filename, hash_dest) == -1) {
		pkg_emit_errno("rename", hash_dest);
		unlink(hash_dest);
		return (EPKG_FATAL);
	}
	if (symlink(hash_dest, filename) == -1) {
		pkg_emit_errno("symlink", hash_dest);
		return (EPKG_FATAL);
	}
	return (EPKG_OK);
}

int
pkg_create_i(struct pkg_create *pc, struct pkg *pkg, bool hash)
{
	struct packing *pkg_archive = NULL;
	int ret;

	unsigned required_flags = PKG_LOAD_DEPS | PKG_LOAD_FILES |
		PKG_LOAD_CATEGORIES | PKG_LOAD_DIRS | PKG_LOAD_SCRIPTS |
		PKG_LOAD_OPTIONS | PKG_LOAD_LICENSES | PKG_LOAD_LUA_SCRIPTS;

	assert(pkg->type == PKG_INSTALLED || pkg->type == PKG_OLD_FILE);

	pkg_archive = pkg_create_archive(pkg, pc, required_flags);
	if (pkg_archive == NULL) {
		if (errno == EEXIST)
			return (EPKG_EXIST);
		pkg_emit_error("unable to create archive");
		return (EPKG_FATAL);
	}

	if ((ret = pkg_create_from_dir(pkg, NULL, pc, pkg_archive, false)) != EPKG_OK) {
		pkg_emit_error("package creation failed");
	}
	packing_finish(pkg_archive);

	if (hash && ret == EPKG_OK)
		ret = hash_file(pkg);

	return (ret);
}

int
pkg_create(struct pkg_create *pc, const char *metadata, const char *plist,
    bool hash)
{
	struct pkg *pkg = NULL;
	struct packing *pkg_archive = NULL;
	int ret = EPKG_FATAL;

	pkg_debug(1, "Creating package");
	if (pkg_new(&pkg, PKG_FILE) != EPKG_OK) {
		return (EPKG_FATAL);
	}

	if (load_metadata(pkg, metadata, plist, pc->rootdir) != EPKG_OK) {
		pkg_free(pkg);
		return (EPKG_FATAL);
	}
	fixup_abi(pkg, pc->rootdir, false);

	pkg_archive = pkg_create_archive(pkg, pc, 0);
	if (pkg_archive == NULL) {
		if (errno == EEXIST) {
			pkg_emit_notice("%s-%s already packaged, skipping...\n",
			    pkg->name, pkg->version);
			pkg_free(pkg);
			return (EPKG_EXIST);
		}
		pkg_free(pkg);
		return (EPKG_FATAL);
	}

	if ((ret = pkg_create_from_dir(pkg, pc->rootdir, pc, pkg_archive, true)) != EPKG_OK)
		pkg_emit_error("package creation failed");

	packing_finish(pkg_archive);
	if (hash && ret == EPKG_OK)
		ret = hash_file(pkg);

	pkg_free(pkg);
	return (ret);
}

static int
pkg_load_message_from_file(int fd, struct pkg *pkg, const char *path)
{
	char *buf = NULL;
	off_t size = 0;
	int ret;
	ucl_object_t *obj;

	assert(pkg != NULL);
	assert(path != NULL);

	if (faccessat(fd, path, F_OK, 0) == -1) {
		return (EPKG_FATAL);
	}

	pkg_debug(1, "Reading message: '%s'", path);
	if ((ret = file_to_bufferat(fd, path, &buf, &size)) != EPKG_OK) {
		return (ret);
	}

	if (*buf == '[') {
		ret = pkg_message_from_str(pkg, buf, size);
		free(buf);
		return (ret);
	}
	obj = ucl_object_fromstring_common(buf, size,
	    UCL_STRING_RAW|UCL_STRING_TRIM);
	ret = pkg_message_from_ucl(pkg, obj);
	ucl_object_unref(obj);
	free(buf);

	return (ret);
}

/* TODO use file descriptor for rootdir */
static int
load_manifest(struct pkg *pkg, const char *metadata, const char *plist,
    const char *rootdir)
{
	int ret;

	ret = pkg_parse_manifest_file(pkg, metadata);

	if (ret == EPKG_OK && plist != NULL)
		ret = ports_parse_plist(pkg, plist, rootdir);
	return (ret);
}

/* TODO use file descriptor for rootdir */
static int
load_metadata(struct pkg *pkg, const char *metadata, const char *plist,
    const char *rootdir)
{
	regex_t preg;
	regmatch_t pmatch[2];
	size_t size;
	int fd, i;

	/* Let's see if we have a directory or a manifest */
	if ((fd = open(metadata, O_DIRECTORY|O_CLOEXEC)) == -1) {
		if (errno == ENOTDIR)
			return (load_manifest(pkg, metadata, plist, rootdir));
		pkg_emit_errno("open", metadata);
		return (EPKG_FATAL);
	}

	if ((pkg_parse_manifest_fileat(fd, pkg, "+MANIFEST")) != EPKG_OK) {
		pkg_emit_error("Error parsing %s/+MANIFEST", metadata);
		close(fd);
		return (EPKG_FATAL);
	}
	/* ensure the uid is properly */
	free(pkg->uid);
	pkg->uid = xstrdup(pkg->name);

	pkg_load_message_from_file(fd, pkg, "+DISPLAY");
	if (pkg->desc == NULL)
		pkg_set_from_fileat(fd, pkg, PKG_ATTR_DESC, "+DESC", false);

	for (i = 0; scripts[i] != NULL; i++) {
		if (faccessat(fd, scripts[i], F_OK, 0) == 0)
			pkg_addscript_fileat(fd, pkg, scripts[i]);
	}

	for (i = 0; lua_scripts[i] != NULL; i++) {
		if (faccessat(fd, lua_scripts[i], F_OK, 0) == 0)
			pkg_addluascript_fileat(fd, pkg, lua_scripts[i]);
	}

	if (plist != NULL && ports_parse_plist(pkg, plist, rootdir) != EPKG_OK) {
		return (EPKG_FATAL);
	}
	close(fd);

	if (pkg->www == NULL) {
		if (pkg->desc == NULL) {
			pkg_emit_error("No www or desc defined in manifest");
			return (EPKG_FATAL);
		}
		regcomp(&preg, "^WWW:[[:space:]]*(.*)$",
		    REG_EXTENDED|REG_ICASE|REG_NEWLINE);
		if (regexec(&preg, pkg->desc, 2, pmatch, 0) == 0) {
			size = pmatch[1].rm_eo - pmatch[1].rm_so;
			pkg->www = xstrndup(&pkg->desc[pmatch[1].rm_so], size);
		} else {
			pkg->www = xstrdup("UNKNOWN");
		}
		regfree(&preg);
	}

	return (EPKG_OK);
}

static void
fixup_abi(struct pkg *pkg, const char *rootdir, bool testing)
{
	bool defaultarch = false;

	/* if no arch autodetermine it */
	if (pkg->abi == NULL) {
		if (ctx.abi.os == PKG_OS_FREEBSD) {
			char *str_osversion;
			xasprintf(&str_osversion, "%d", pkg_abi_get_freebsd_osversion(&ctx.abi));
			pkg_kv_add(&pkg->annotations, "FreeBSD_version", str_osversion, "annotation");
		}
		pkg->abi = pkg_abi_to_string(&ctx.abi);
		defaultarch = true;
	}

	if (!testing)
		pkg_analyse_files(NULL, pkg, rootdir);

	if (ctx.developer_mode)
		suggest_arch(pkg, defaultarch);
}

int
pkg_load_metadata(struct pkg *pkg, const char *mfile, const char *md_dir,
    const char *plist, const char *rootdir, bool testing)
{
	int ret;

	ret = load_metadata(pkg, md_dir != NULL ? md_dir: mfile, plist, rootdir);
	if (ret != EPKG_OK)
		return (ret);

	fixup_abi(pkg, rootdir, testing);
	return (ret);
}

static int64_t	count;
static int64_t  maxcount;
static const char *what;

static int magnitude(int64_t num)
{
	int oom;

	if (num == 0)
		return (1);
	if (num < 0)
		num = -num;

	for (oom = 1; num >= 10; oom++)
		num /= 10;

	return (oom);
}

static void
counter_init(const char *count_what, int64_t max)
{
	count = 0;
	what = count_what;
	maxcount = max;
	pkg_emit_progress_start("%-20s%*s[%jd]", what,
	    6 - magnitude(maxcount), " ", (intmax_t)maxcount);

	return;
}

static void
counter_count(void)
{
	count++;

	if (count % TICK == 0)
		pkg_emit_progress_tick(count, maxcount);

	return;
}

static void
counter_end(void)
{
	pkg_emit_progress_tick(count, maxcount);
	return;
}
